### PR TITLE
Add `UnaryServerCall`

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

`ExchangeType` is introduced by #3956. A unary call is automatically
aggregated by `Http{1,2}RequestDecoder` and `HttpServerHandler`.
The optimized call path for non-stream requests showed better
performance. As a part of the performance of optimizations, we can add a
specialized `ServerCall` implementation for unary calls. Because
unary calls are widely used for general communication.
Even if the current implementation is not a significant improvement, we
can add further optimization based on this.

Modifications:

- Split `ArmeriaServerCall` into `UnaryServerCall` and
  `StreamingServerCall`.
  - `AbstractServerCall` is added to share common logic.
- Add `UnaryDecoderInput` as a variant of `StreamDecoderInput` which
  holds a single `ByteBuf`.
  - Renamed `ByteBufDecoderInput` into `ByteBufsDecoderInput` to
    differenciate from `UnaryDecoderInput`
- Add `UnaryMessageDeframer` that directly decodes an `HttpData` into
  `DeframedMessage`.
- Miscellaneous) Remove an unexpected string concatenation when validating
  `ResponseHeaders` in `RegularFixedHttpResponse`

Result:

Better performance for unary gRPC call.

### CODE REVIEW

**Comment by ikhoon** [comment]
Benchmark results:

|Branch  |Benchmark                   | Mode  |Cnt  |     Score |     Error  |Units|
|--------|----------------------------|-------|-----|-----------|------------|-----|
|current |GrpcServiceBenchmark.empty  |thrpt  |  5  |370008.598 |± 5016.314  |ops/s|
|master  |GrpcServiceBenchmark.empty  |thrpt  |  5  |347432.389 |± 12037.618 |ops/s|

File: core/src/main/java/com/linecorp/armeria/common/FixedHttpResponse.java
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/FixedHttpResponse.java:61]: 😱

File: core/src/main/java/com/linecorp/armeria/internal/common/stream/ByteBufsDecoderInput.java
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/internal/common/stream/ByteBufsDecoderInput.java:30]: :+1:

File: grpc-protocol/src/main/java/com/linecorp/armeria/common/grpc/protocol/AbstractMessageDeframer.java
Comment (trustin) [inline grpc-protocol/src/main/java/com/linecorp/armeria/common/grpc/protocol/AbstractMessageDeframer.java:34]:
```suggestion
 * A skeletal implementation of gRPC message deframer. See
```

File: grpc/src/main/java/com/linecorp/armeria/server/grpc/FramedGrpcService.java
Comment (trustin) [inline grpc/src/main/java/com/linecorp/armeria/server/grpc/FramedGrpcService.java:353]: Not probably worth doing this because we down-cast below anyway.

File: grpc/src/main/java/com/linecorp/armeria/server/grpc/UnaryServerCall.java
Comment (trustin) [inline grpc/src/main/java/com/linecorp/armeria/server/grpc/UnaryServerCall.java:118]: Would it be better collecting the received `HttpData` instances as they are rather than merging them? If we received only one `HttpData` (excluding the empty ones), we could use `UnaryMessageDeframer` but otherwise we could use less optimized one.

File: grpc/src/main/java/com/linecorp/armeria/server/grpc/AbstractServerCall.java
Comment (trustin) [inline grpc/src/main/java/com/linecorp/armeria/server/grpc/AbstractServerCall.java:112]: We might want to merge @ta7uw's autoCompression PR before merging this in.

File: grpc/src/main/java/com/linecorp/armeria/server/grpc/AbstractServerCall.java
Comment (trustin) [inline grpc/src/main/java/com/linecorp/armeria/server/grpc/AbstractServerCall.java:218]: Probably fine to rename to `close()` because there's no signature conflict.

File: grpc/src/main/java/com/linecorp/armeria/server/grpc/AbstractServerCall.java
Comment (ikhoon) [reply grpc/src/main/java/com/linecorp/armeria/server/grpc/AbstractServerCall.java:112]: Thanks for the ping. 😀

File: grpc/src/main/java/com/linecorp/armeria/server/grpc/AbstractServerCall.java
Comment (ikhoon) [inline grpc/src/main/java/com/linecorp/armeria/server/grpc/AbstractServerCall.java:473]: This part was slightly changed from `setCompression(encoding)` while the master branch was merged.
`setCompression(encoding)` calls `responseFramer.setCompressor(ForwardingCompressor.forGrpc(compressor))` 
that is also invoked the end of this block.
/cc @ta7uw

Comment (ikhoon) [reply grpc/src/main/java/com/linecorp/armeria/server/grpc/AbstractServerCall.java:473]: https://github.com/line/armeria/blob/1008ba517cf7dd9df5699d0ca62e3f451464d66c/grpc/src/main/java/com/linecorp/armeria/server/grpc/ArmeriaServerCall.java#L267

**Review by minwoox — COMMENTED** [review]
Left some minor comments. Looks really good 👍

File: grpc/src/main/java/com/linecorp/armeria/server/grpc/FramedGrpcService.java
Comment (ikhoon) [reply grpc/src/main/java/com/linecorp/armeria/server/grpc/FramedGrpcService.java:262]: Ignore the comment avoid. 
As `UnaryServerCall` provides an optimized implementation for unary calls, `CLIENT_STREAMING` is still handled by `StreamingServerCall`.
This condition is added to inject an `HttpResponse` optimized for a non-streaming response.
We may provide a lighter implementation rather than a combination of a fixed `HttpResponse` and deferred `HttpResponse`.

File: grpc/src/main/java/com/linecorp/armeria/server/grpc/UnaryServerCall.java
Comment (ikhoon) [reply grpc/src/main/java/com/linecorp/armeria/server/grpc/UnaryServerCall.java:103]: I will go with the latter one.

Comment (ikhoon) [reply grpc/src/main/java/com/linecorp/armeria/server/grpc/UnaryServerCall.java:103]: On second thought, renaming looks better because `startDeframing()` is explicitly called when a `listener` is set for which nullness of `listener` need to check.

File: grpc/src/main/java/com/linecorp/armeria/server/grpc/UnaryServerCall.java
Comment (ikhoon) [reply grpc/src/main/java/com/linecorp/armeria/server/grpc/UnaryServerCall.java:125]: We don't need to release the data here.
Because an exception is raised when a deframer detects malformed data
which is released by the try-with-resource block. 
Anyway, will leave an explanation for that.

Comment (ikhoon) [reply grpc/src/main/java/com/linecorp/armeria/server/grpc/UnaryServerCall.java:125]: https://github.com/line/armeria/blob/0c5692de73b005cb8a124278ab37bf56b7bd3b99/grpc/src/main/java/com/linecorp/armeria/server/grpc/UnaryMessageDeframer.java#L64

File: grpc/src/main/java/com/linecorp/armeria/server/grpc/UnaryServerCall.java
Comment (ikhoon) [reply grpc/src/main/java/com/linecorp/armeria/server/grpc/UnaryServerCall.java:206]: Nice idea. It looks better to use `finally` block to call `closeListener`.

**Review by minwoox — APPROVED** [review]
LGTM once https://github.com/line/armeria/pull/4192#discussion_r896409595 is addressed. Thanks!

**Review by jrhee17 — APPROVED** [review]
Looks great! Thanks @ikhoon 👍 🙇 🚀

File: grpc/src/main/java/com/linecorp/armeria/server/grpc/AbstractServerCall.java
Comment (ta7uw) [reply grpc/src/main/java/com/linecorp/armeria/server/grpc/AbstractServerCall.java:473]: I checked the change of this part.
Looks good!

File: grpc/src/test/java/com/linecorp/armeria/client/grpc/GrpcClientCompressionTest.java
Comment (trustin) [inline grpc/src/test/java/com/linecorp/armeria/client/grpc/GrpcClientCompressionTest.java:52]: Should we now test streaming requests and responses here as well?

**Review by trustin — COMMENTED** [review]
Left super minor comments 😅

File: grpc/src/test/java/com/linecorp/armeria/client/grpc/GrpcClientCompressionTest.java
Comment (ikhoon) [reply grpc/src/test/java/com/linecorp/armeria/client/grpc/GrpcClientCompressionTest.java:52]: The logic is still handled in one place, `AbstractServerCall.doSendHeaders()`, but it would be good to add tests just in case.

File: grpc/src/test/java/com/linecorp/armeria/server/grpc/UnaryServerCallTest.java
Comment (ikhoon) [reply grpc/src/test/java/com/linecorp/armeria/server/grpc/UnaryServerCallTest.java:91]: Makes sense!

File: grpc/src/test/java/com/linecorp/armeria/client/grpc/GrpcClientCompressionTest.java
Comment (ikhoon) [reply grpc/src/test/java/com/linecorp/armeria/client/grpc/GrpcClientCompressionTest.java:60]: Added at [`dcae4a7` (#4192)](https://github.com/line/armeria/pull/4192/commits/dcae4a747ff83c209af08e20201a346976e17238)

**Pull request merged**
